### DISCUSSION TITLE
How to process a psd image and keep the alpha but not apply transparency

### DISCUSSION THREAD

**Discussion by pixelvapour** [ImageMagick/ImageMagick#6477] [Help]
### ImageMagick version

7.1.1-11 Q16-HDRI x86_64 21206

### Operating system, version and so on

MacOS Ventura 13.3.1 (a) 

### Description

I have many CMYK and RGB images that require resizing. A percentage of these images contain an alpha channel. The alpha channel needs to remain in the PSD for the occasions when the original images background needs replacing.
The issue I'm having is that the output resized image in psd format, gives me either black background or a transparent background with all the different settings I have tried. I've also had trouble getting the psd output to be flattened, so Photoshop shows 'Background' in the layers palette. 
I've been successful doing this with tif format as the input and output, but not with psd.

Here is the command I've used for tif.
magick RGB-with_alpha.tif -background None -compose Copy -resize 50% -define tiff:alpha=unspecified IM_resized_alpha.tif

I'd be OK with the final output image being tif but I'd need some way of reading in the psd, converting it to tif, reading it back in and then processing it for scale and keeping the all important alpha channel.

I have searched over quite a few days to find a solution and down a few rabbit holes, but if I've missed something obvious the I apologise in advance.

**Comment by snibgo** [comment]
Please link to your input PSD file.

**Reply by dlemstra** [reply]
Or even better upload it here zipped.

**Comment by pixelvapour** [comment]
Thanks for looking at my query. Here's an image in both PSD and an LZW compressed TIF.
[SV22HEPA_FILTER_067-RGB-3QLeft-A5_WS.tif.zip](https://github.com/ImageMagick/ImageMagick/files/12024493/SV22HEPA_FILTER_067-RGB-3QLeft-A5_WS.tif.zip)
[SV22HEPA_FILTER_067-RGB-3QLeft-A5_WS.psd.zip](https://github.com/ImageMagick/ImageMagick/files/12024494/SV22HEPA_FILTER_067-RGB-3QLeft-A5_WS.psd.zip)

**Comment by snibgo** [comment]
I suggest inserting "`-alpha discrete`". This retains the whitish background, and an "unspecified" alpha channel with varying values.

The input PSD contains negative values. If you want to retain these, I suggest inserting "`-define quantum:format=floating-point -depth 32`". This will save the TIFF as 32-bit/channel floating point. If you want floating-point but clamped, also insert `-clamp`.

```
magick SV22HEPA_FILTER_067-RGB-3QLeft-A5_WS.psd -define quantum:format=floating-point -depth 32 -alpha discrete -define tiff:alpha=unspecified x.tiff
```

**Comment by pixelvapour** [comment]
Thanks for the command snibgo.
I'm not sure where the 32bit data is coming from but I definitely don't need it in the final output. When I tried the full command it generate a very washed out image.
When I removed -define quantum:format=floating-point -depth 32 but retained -alpha discrete the original colour of the image was correct but there's a black aliased edge around RGB image, which isn't present on the alpha.
Is there a way to remove this?
<img width="1999" alt="Screenshot 2023-07-13 at 09 40 55" src="https://github.com/ImageMagick/ImageMagick/assets/668745/8d2555f1-ccc1-4f4a-8319-406107976108">

**Comment by pixelvapour** [comment]
I have an alternative route but I'd need a way to get this command working to give me a TIF image which looks identical to the PSD, which I'm unable to do due to the way the alpha is treated. Is is possible? The example TIF I uploaded was saved out of Photoshop.
magick /SV22HEPA_FILTER_067-RGB-3QLeft-A5_WS.psd  /xx.tif

I've tried all of the -alpha processing options but none have been successful or me.
I'm sure it's possible but my IM knowledge is lacking at this stage.

**Comment by snibgo** [comment]
Some image viewers mis-read 32-bit floating point TIFF files, making the colour look "washed-out". You might try 16-bit.

For the thin black line, insert "`-define psd:alpha-unblend=off`" before reading the PSD. See https://imagemagick.org/script/defines.php
```
magick -define psd:alpha-unblend=off SV22HEPA_FILTER_067-RGB-3QLeft-A5_WS.psd -define quantum:format=floating-point -depth 32 -alpha discrete -define tiff:alpha=unspecified x.tiff
```

**Comment by fmw42** [comment]
Using my Photoshop on my Mac, I can see that the resize works fine for TIFF but not for PSD.  The PSD loses the white layer background, which becomes black.  (I have no problem with the alpha channel not being preserved!)  However, when comparing the results for TIFF and PSD, I do not see anything in my Photoshop that shows that background color, let alone does not let me change it.  So I think this is buried in the PSD profiles somewhere.  I think this is a bug in IM handling the resize of PSD. So I think an IM developer (Dirk) will likely need to look into this.

**Comment by snibgo** [comment]
We can use IM to see the object and the background colour with:
```
magick -define psd:alpha-unblend=off SV22HEPA_FILTER_067-RGB-3QLeft-A5_WS.psd -alpha off x.png
```

**Comment by pixelvapour** [comment]
Fantastic. Snibgo, the command you supplied works as I'd hoped.
magick -define psd:alpha-unblend=off SV22HEPA_FILTER_067-RGB-3QLeft-A5_WS.psd -alpha discrete -define tiff:alpha=unspecified  -compress LZW Output_with_alpha_LZW_compressed.tif

This takes a PSD input which contains an alpha channel and outputs an LZW compressed TIF which retains the alpha and crucially, keeps the original PSD's 'background' untouched. Perfect.
Thanks so much for your help as I'd not have arrived at this solution without your input. I also feel better equipped to tackle similar issues going forward.
